### DISCUSSION TITLE
Device detection library: Should we replace UA parser?

### DISCUSSION THREAD

**Discussion by mposolda** [keycloak/keycloak#17067] [Feedback]
Right now, we use UA parser as a library for "device detection". For example Keycloak account console uses this to detect what was the device used to login the particular browser session. It uses mostly the `User-Agent` HTTP header sent by the browser and tries to detect the device based on it.

UA parser is 3d party library, but it does not seem to be maintained. So we're looking for an alternative to it. Right now, we have prototype PR [1], which replaces UA parser with YAUAA [2]. Does anyone have any experience with YAUAA? Or suggestions to other alternatives?

[1] https://github.com/keycloak/keycloak/pull/16876/files
[2] https://yauaa.basjes.nl/

**Comment by douglaspalmer** [comment]
I wrote a test app today to test YAUAA vs UA-Parser today. In doing so I discovered that UA-Parser has a cached version of their parser and they did a new release 4 days ago. My test app timed the parser init; then it timed parsing the 56 most common user agents; then it timed parsing those user agents again 10,000 times. Here are the results in milliseconds:

YAUAA init took: 1700
Initial 56 agent YAUAA parse took: 946
10000x56 YAUAA cached parses took: 1215

YAUAA uncached init took: 1693
Initial 56 agent YAUAA uncached parse took: 958
10000x56 YAUAA uncached parses took: 128087

UACachingParser init took: 210
Initial 56 agent UACachingParser parse took: 140
10000x56 UACachingParser cached parses took: 25

UAParser init took: 224
Initial 56 agent UAParser parse took: 120
10000x56 UAParser uncached parses took: 741317

**Comment by douglaspalmer** [comment]
UA parser had a release this week. The project isn't very active but it isn't dead either. I think we should switch to their caching parser and keep an eye on the project.

**Comment by abstractj** [comment]
@mposolda @douglaspalmer my major concern is about the level of maintenance of the project when it comes to releases and fixing CVEs. They got a new release this week, it appears that a single person is responsible for it, which is risky. Even if we contribute with bug fixes or updating dependencies, we still depend on a single maintainer to release a new version.

A concrete example of what I say is the issue created on Oct, 2022 - https://github.com/ua-parser/uap-java/issues/74

**Reply by mposolda** [reply]
@abstractj I agree that having project dependent on single maintainer is not ideal. On the other hand, the yauaa has same issue (single maintainer, even if very active at this moment) and we did not found anything better as open-source library. Is it better @douglaspalmer ?

So right now, my vote is to also go with UACachingParser (so stick to ua-parser, but update to latest version and update it to use "caching").

I prefer ua-parser instead of yauaa right now due the performance concerns. But we can possibly revisit in the future.

**Comment by stianst** [comment]
+1 to stick with ua parser. I assume there's also not the big memory penalty with ua parser either?

End of the day we could even fork the java side of ua parser as that's pretty small and simple. Main part is in the regex, which has very low likelihood of having any cves

**Comment by nielsbasjes** [comment]
I'm totally cool with you making the choice to stay with the analyzer you have right now.
What puzzles me is the argument of performance.

- In the uncached case the yauaa is more than 5 times faster than ua-parser.
- In the cached case the cache used by ua-parser is a lot (like really a lot) faster than the cache used by Yauaa.

I had a quick look at the ua-parser code how they achieved this and the reason for this is quite simple: 
- Yauaa uses https://github.com/ben-manes/caffeine for caching which is a very fast **threadsafe** implementation.
- The ua-parser uses an unsynchronized LRUMap which is a lot faster because it is **not threadsafe**. 

https://github.com/ua-parser/uap-java/blob/50535dbc02953ccf4614006e26ace52af1c3b8b4/src/main/java/ua_parser/CachingParser.java#LL66C25-L66C31

I also think that some of the testvalues (for both) are a bit skewed because of a `cold` JVM.

**Reply by nielsbasjes** [reply]
and Please ... do NOT check who the original author is of this CachingParser ...

**Reply by nielsbasjes** [reply]
If you really need the **unsychronized** LRUMap speed you can override the caching implementation in Yauaa.

Have a look at this test as an example:
https://github.com/nielsbasjes/yauaa/blob/main/analyzer/src/test/java/nl/basjes/parse/useragent/TestCaching.java#L246

```java
UserAgentAnalyzer uaa = UserAgentAnalyzer
    .newBuilder()
    .withCache(10000)
    .withCacheInstantiator((CacheInstantiator) LRUMap::new)
    .withClientHintsCache(10000)
    .withClientHintCacheInstantiator((ClientHintsCacheInstantiator<Serializable>) LRUMap::new)
    .build();
```

**Reply by nielsbasjes** [reply]
FYI: @mposolda @douglaspalmer @abstractj @stianst

**Reply by douglaspalmer** [reply]
@nielsbasjes I did a quick test using YAUAA and LRUMap as you suggested. It actually came out slower than YAUAA with Caffeine

YAUAA LRUMap init took: 1649
Initial YAUAA LRUMAP parse took: 1115
10000 YAUAA LRUMAP cached parses took: 1220

**Reply by nielsbasjes** [reply]
That is unexpected.
At least be aware the cache you have now is not threadsafe.
